README.md
# TEXTA Toolkit 3

## Documentation

https://docs.texta.ee

## Wiki

https://git.texta.ee/texta/texta-rest/wikis/home

## Notes

Works with Python 3.8

### Creating environment:

`conda env create -f environment.yaml`

Running migrations:

`python3 migrate.py`

* This script will also create an admin account with the default username "admin". You can
  use ```python migrate.py -u {{username}}``` instead for a custom username of your choice.
* Password for that admin account will be generated automatically and printed to the console. This behaviour can be
  overwritten with the environment variable `TEXTA_ADMIN_PASSWORD`, in which case the password will be set to the same
  value as 'TEXTA_ADMIN_PASSWORD'.
* Running ```python migrate.py -o``` will overwrite the password with whatever value you have inside the environment
  variable `TEXTA_ADMIN_PASSWORD`.

Running application:

`python3 manage.py runserver`

`celery -A toolkit.taskman worker -l info`

Import testing data:

`python3 import_test_data.py`

Run all tests:

`python3 manage.py test`

Run tests for specific app:

`python3 manage.py test appname (eg python3 manage.py test toolkit.neurotagger)`

Run performance tests (not run by default as they are slow):

`python3 manage.py test toolkit.performance_tests`

Building Docker:

`docker build -t texta-rest:latest -f docker/Dockerfile .`

Running Docker:

`docker run -p 8000:8000 texta-rest:latest`

Building Docker with GPU support:

`docker build -t texta-rest:gpu-latest -f docker/gpu.Dockerfile .`

Running Docker with GPU support requires NVIDIA Container Toolkit to be installed on the host
machine: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker. When Container
Toolkit is installed:

`docker run --gpus all -p 8000:8000 texta-rest:latest-gpu`

# Environment variables

## Deploy & Testing variables

* TEXTA_ENV_FILE - Optional file path for a typical .env file to be loaded into memory for TEXTA Toolkit (Default: None).
* TEXTA_SECRET_KEY - ***String key for cryptographic security purposes. ALWAYS SET IN PRODUCTION.***

* TEXTA_CORS_ALLOW_CREDENTIALS - Whether to allow cookies to be included in cross-site HTTP requests (Default: True)
* TEXTA_CORS_ALLOW_ALL_ORIGINS - Whether to allow requests from all Origins (Default: false)

* TEXTA_CELERY_USED_QUEUES - Comma separated list of Celery queues you are using for TEXTA Toolkit. No need to touch when running a standard configuration.
* TEXTA_ELASTIC_VERSION - Must equal to the integer of the main Elasticsearch cluster version (Default: 6).
* TEXTA_DEPLOY_KEY - Used to separate different Toolkit instances for cases where Elasticsearch or the database are
  shared amongst multiple instances. Best to give this a simple number (Default: 1).
* TEXTA_ADMIN_PASSWORD - Password of the admin user created on first run.
* TEXTA_USE_CSRF - Whether to disable CSRF for integration tests (Default: false).
* TEXTA_CELERY_ALWAYS_EAGER - Whether to use Celerys async features or not, useful for testing purposes locally. (
  Default: False)


* TEXTA_DATA_DIR - Path to the directory in which TEXTA Toolkit saves the models it generates, and the binary model
  dependencies it needs (Default: data).
* TEXTA_EXTERNAL_DATA_DIR - Path to the base directory in which 3rd party models (MLP/BERT/etc) are kept (Default:
  data/models).
* TEXTA_CACHE_DIR - Path for the cache folder which BERT uses (Default: data/external/.cache).
* TEXTA_RELATIVE_MODELS_DIR - Relative path of the directory in which all the different types of models are stored in.
  (Default: "/data/models").
* TEXTA_LANGUAGE_CODES - Comma separated string of Stanza supported language codes to use for Multilingual Processing.
  (Default: "").
* TEXTA_MLP_USE_GPU - Use GPU to speed up MLP (Default: False).
* TEXTA_MLP_MODEL_DIRECTORY_PATH - Relative path to the directory into which Stanza models will be stored under the "
  stanza" folder (setting this to ./home/texta will create ./home/texta/stanza which contains subfolders for every
  language like ./home/texta/stanza/et etc). (Default: "./data/external/mlp").
* TEXTA_MLP_DEFAULT_LANGUAGE - Language code of the language the MLP module will default to when trying to analyze a document whichs language it could not detect properly (Default: en).
* TEXTA_ALLOW_BERT_MODEL_DOWNLOADS - Boolean flag indicating if the users can download additional BERT models.
  (Default: True).
* TEXTA_BERT_MODEL_DIRECTORY_PATH - Relative path to the directory into which pretrained and fine-tuned BERT models will
  be stored under the "bert_tagger" folder. (setting this to ./home/texta will create
  ./home/texta/bert_tagger/pretrained/ which contains subfolders for every downloaded bert more like
  ./home/texta/bert_model/pretrained/bert-base-multilingual-cased etc and ./home/texta/bert_model/fine_tuned/ which will
  store fine-tuned BERT models. (Default: "./data/models").
* TEXTA_NLTK_DATA_DIRECTORY_PATH - Path of the directory where the NLTK library keeps its resources (Default:
  data/external/nltk).
* TEXTA_BERT_MODELS - Comma seprated string of pretrained BERT models to download.
  (Default: "bert-base-multilingual-cased,bert-base-uncased,EMBEDDIA/finest-bert").
* SKIP_BERT_RESOURCES - If set "True", skips downloading pretrained BERT models. (Default: false).
* SKIP_MLP_RESOURCES - Whether to skip downloading MLP resources on application boot-up (Default: false).
* SKIP_NLTK_RESOURCES - Whether to skip downloading NLTK library resources on application boot-up (Default: false).
* TEXTA_EVALUATOR_MEMORY_BUFFER_GB - The minimum amount of memory that should be left free while using the evaluator,
  unit = GB. (Default = 50% of available_memory)
* TEXTA_DATASOURCE_CHOICES - Choices for index domain field given as a list ex: [['prefix_name', 'display_name']]. (
  Default = [["emails", "emails"], ["news articles", "news articles"], ["comments", "comments"]
  , ["court decisions", "court decisions"], ["tweets", "tweets"], ["forum posts", "forum posts"]
  , ["formal documents", "formal documents"], ["other", "other"]])

* TOOLKIT_PROJECT_DATA_PATH - Path of the directory in which project specific data is kept (Default: data/projects).

## External services
* TEXTA_ES_PREFIX - String used to limit Elasticsearch index access. Only indices matched will be the ones matching "
  {TEXTA_ES_PREFIX}*".
* TEXTA_ES_URL - URL of the Elasticsearch instance including the protocol, host and port (ex. http://localhost:9200).
* TEXTA_REDIS_URL - URL of the Redis instance including the protocol, host and port (ex. redis://localhost:6379).

## Django specifics

* TEXTA_CORS_ORIGIN_WHITELIST - Comma separated string of urls (**NO WHITESPACE**) for the CORS whitelist. Needs to
  include the protocol (ex. http://* or http://*,http://localhost:4200).
* TEXTA_ALLOWED_HOSTS - Comma separated string (**NO WHITESPACE**) representing the host/domain names that this Django
  site can serve (ex. * or *,http://localhost:4200).
* TEXTA_DEBUG - True/False values on whether to run Django in its Debug mode or not (Default: true).
* TEXTA_MAX_UPLOAD - Maximum size of files in bytes that are allowed to be updated, which Django validates (Default: 1073741824 aka 1GB)

## Database credentials

* DJANGO_DATABASE_ENGINE - https://docs.djangoproject.com/en/3.0/ref/settings/#engine
* DJANGO_DATABASE_NAME - The name of the database to use. For SQLite, it’s the full path to the database file. When
  specifying the path, always use forward slashes, even on Windows.
* DJANGO_DATABASE_USER - The username to use when connecting to the database. Not used with SQLite.
* DJANGO_DATABASE_PASSWORD - The password to use when connecting to the database. Not used with SQLite.
* DJANGO_DATABASE_HOST - Which host to use when connecting to the database. An empty string means localhost. Not used
  with SQLite.
* DJANGO_DATABASE_PORT - The port to use when connecting to the database. An empty string means the default port. Not
  used with SQLite.

## Docker specific configurations:
* TEXTA_SHORT_TASK_WORKERS - Number of processes available for short term tasks (Default: 2).
* TEXTA_LONG_TASK_WORKERS - Number of processes available for long term tasks (Default: 4).
* TEXTA_MLP_TASK_WORKERS - Number of processes available for MLP based tasks (Default: 2).
* TEXTA_SHORT_MAX_TASKS - Number of tasks per worker for short term tasks (Default: 10).
* TEXTA_LONG_MAX_TASKS - Number of tasks per worker for long term tasks (Default: 10).
* TEXTA_MLP_MAX_TASKS - Number of tasks per worker for MLP based tasks (Default: 10).
* TEXTA_BEAT_LOG_LEVEL - Which log level should beat output within the Docker image (Default: WARNING).
* TEXTA_CELERY_LOG_LEVEL - Which log level should Celery workers output within the Docker image (Default: WARNING).



## Extra Elasticsearch connection configurations

Unless you have a specially configured Elasticsearch instance, you can ignore these options.

* TEXTA_ES_USER - Username to authenticate to a secured Elasticsearch instance.
* TEXTA_ES_PASSWORD - Password to authenticate to a secured Elasticsearch instance.

https://elasticsearch-py.readthedocs.io/en/6.3.1/connection.html#elasticsearch.Urllib3HttpConnection:

* TEXTA_ES_USE_SSL
* TEXTA_ES_VERIFY_CERTS
* TEXTA_ES_CA_CERT_PATH
* TEXTA_ES_CLIENT_CERT_PATH
* TEXTA_ES_CLIENT_KEY_PATH
* TEXTA_ES_TIMEOUT
* TEXTA_ES_SNIFF_ON_START
* TEXTA_ES_SNIFF_ON_FAIL

## UAA specific configurations

* TEXTA_USE_UAA - Whether to include UAA authentication with the default authentication (Default: false).
* TEXTA_UAA_SCOPES - Which scopes should be sent with communication between TEXTA Toolkit and UAA (Default: openid
  texta.*).
* TEXTA_UAA_SUPERUSER_SCOPE - Which scope to use for determining whether an UAA user is a superuser (Default:
  texta.admin).
* TEXTA_UAA_PROJECT_ADMIN_SCOPE - Which scope to use to specify whether an UAA user has project administrator rights to
  ANY project which is available to them (Default: texta.project_admin).
* TEXTA_UAA_SCOPE_PREFIX - Prefix for determining UAA user access to TEXTA Toolkit. Any user who does not have a scope
  which matches the pattern "{TEXTA_UAA_SCOPE_PREFIX}.*" will be denied entry to TEXTA Toolkit (Default: texta).

* TEXTA_UAA_URL - URI for the UAA service (Default: http://localhost:8080).
* TEXTA_UAA_REDIRECT_URI - URI into which the user will be redirected after a successful UAA login (
  Default: http://localhost:8000/api/v2/uaa/callback).
* TEXTA_UAA_FRONT_REDIRECT_URL - Configuration for the front end to determine where it will redirect the user after a
  successful login with UAA by Toolkit (Default: http://localhost:4200/oauth/uaa)
* TEXTA_UAA_CLIENT_ID - UAA client ID for authenticating the TEXTA Toolkit application for UAA. ***Must be kept
  secret***
* TEXTA_UAA_CLIENT_SECRET - Password for authenticating the TEXTA Toolkit application with UAA. ***Must be kept
  secret***


CHANGELOG.md
## 3.7.9 (2022-11-14)

### Features (2 changes)

- [feature: Added support for clearing read_only blocks on the cluster.](texta/texta-rest@f5c8eff4e6658dd31ac50bedfdde04797e8384d7) by @mkollo
 ([merge request](texta/texta-rest!170))
- [feature: Added S3 support to taggers and bert taggers.](texta/texta-rest@6a49e91c157db61a3d5a4d5643db49b00e8f80f3) by @mkollo
 ([merge request](texta/texta-rest!169))

### Updates (1 change)

- [update: Added S3 support for importing and exporting Tagger Groups from S3](texta/texta-rest@a7d1ab4724bb58cd01fb42f85604aa7f9ee13d19) by @mkollo

## 3.7.8 (2022-11-10)

### Features (1 change)

- [feature: Added S3 support to taggers and bert taggers.](texta/texta-rest@6a49e91c157db61a3d5a4d5643db49b00e8f80f3) by @mkollo
 ([merge request](texta/texta-rest!169))

### Updates (1 change)

- [update: Added S3 support for importing and exporting Tagger Groups from S3](texta/texta-rest@a7d1ab4724bb58cd01fb42f85604aa7f9ee13d19) by @mkollo

## 3.7.7 (2022-11-07)

### Features (1 change)

- [feature: Added S3 support to taggers and bert taggers.](texta/texta-rest@6a49e91c157db61a3d5a4d5643db49b00e8f80f3) by @mkollo
 ([merge request](texta/texta-rest!169))

## 3.7.6 (2022-10-26)

### Updates (1 change)

- [update: Added support for changing default language code for MLP through env variables.](texta/texta-rest@894b3cf9fb6c036f416595901efd3a399b35ccd2) by @mkollo

## 3.7.5 (2022-10-21)

No changes.

## 3.7.4 (2022-09-27)

### Updates (1 change)

- [update: VERSION 3.7.4](texta/texta-rest@14058a454a99e7a1ee8f330ec6e4f7a4a9ff4cca) by @wael

## 3.7.3 (2022-09-27)

### Bug fixes (1 change)

- [bug: Fixed issue of bigger TaggerGroups failing on creating due to timeouts.](texta/texta-rest@33243750ffebce616f07251cbfe7e75fad0cdd76) by @mkollo
 ([merge request](texta/texta-rest!167))

### Updates (4 changes)

- [update: Added configuration for Bert Taggers to use a dynamically configurable different queue.](texta/texta-rest@8cdf1acf7b2e95fd34feacf1a976c2a715d91997) by @mkollo
 ([merge request](texta/texta-rest!164))
- [update: Storing classes inside the evaluator model.](texta/texta-rest@5813e650b3c3c03e66682bfcab81872c14d6faa7) by @mkollo
 ([merge request](texta/texta-rest!165))
- [update: README TTK v3](texta/texta-rest@f84d0cb98b544b3a93386c3272c4c4aeec3f1000) by @wael

- [update: Added ability to favorite tasks to Evaluators.](texta/texta-rest@199739240e08fdeb9ec8ba4f017e1ecb1c3040ea) by @mkollo
 ([merge request](texta/texta-rest!163))

## 3.7.2 (2022-09-08)

### Updates (4 changes)

- [update: Added configuration for Bert Taggers to use a dynamically configurable different queue.](texta/texta-rest@8cdf1acf7b2e95fd34feacf1a976c2a715d91997) by @mkollo
 ([merge request](texta/texta-rest!164))
- [update: Storing classes inside the evaluator model.](texta/texta-rest@5813e650b3c3c03e66682bfcab81872c14d6faa7) by @mkollo
 ([merge request](texta/texta-rest!165))
- [update: README TTK v3](texta/texta-rest@f84d0cb98b544b3a93386c3272c4c4aeec3f1000) by @wael

- [update: Added ability to favorite tasks to Evaluators.](texta/texta-rest@199739240e08fdeb9ec8ba4f017e1ecb1c3040ea) by @mkollo
 ([merge request](texta/texta-rest!163))

## 3.7.1 (2022-09-07)

### Updates (4 changes)

- [update: Added configuration for Bert Taggers to use a dynamically configurable different queue.](texta/texta-rest@8cdf1acf7b2e95fd34feacf1a976c2a715d91997) by @mkollo
 ([merge request](texta/texta-rest!164))
- [update: Storing classes inside the evaluator model.](texta/texta-rest@5813e650b3c3c03e66682bfcab81872c14d6faa7) by @mkollo
 ([merge request](texta/texta-rest!165))
- [update: README TTK v3](texta/texta-rest@f84d0cb98b544b3a93386c3272c4c4aeec3f1000) by @wael

- [update: Added ability to favorite tasks to Evaluators.](texta/texta-rest@199739240e08fdeb9ec8ba4f017e1ecb1c3040ea) by @mkollo
 ([merge request](texta/texta-rest!163))

## 3.7.0 (2022-09-07)

### Updates (3 changes)

- [update: Storing classes inside the evaluator model.](texta/texta-rest@5813e650b3c3c03e66682bfcab81872c14d6faa7) by @mkollo
 ([merge request](texta/texta-rest!165))
- [update: README TTK v3](texta/texta-rest@f84d0cb98b544b3a93386c3272c4c4aeec3f1000) by @wael

- [update: Added ability to favorite tasks to Evaluators.](texta/texta-rest@199739240e08fdeb9ec8ba4f017e1ecb1c3040ea) by @mkollo
 ([merge request](texta/texta-rest!163))

## 3.6.0 (2022-09-06)

### Updates (1 change)

- [update: Added ability to favorite tasks to Evaluators.](texta/texta-rest@199739240e08fdeb9ec8ba4f017e1ecb1c3040ea) by @mkollo
 ([merge request](texta/texta-rest!163))

## 3.5.3 (2022-09-05)

### Features (2 changes)

- [feat: Reformat task objects to allow multiple task records + normalization of common fields.](texta/texta-rest@d2bf6d6f516503482d3655be9192bb3c1524a902) by @mkollo
 ([merge request](texta/texta-rest!158))
- [feat: Automatically add index author to procedures that generate a new index.](texta/texta-rest@e9bcdcdeb109b9ba9c7cc03da55061e8ab7a3126) by @mkollo
 ([merge request](texta/texta-rest!157))

### Bug fixes (2 changes)

- [bug: Fixed searcher export not working bc of authentication issues.](texta/texta-rest@b8ae7c8c20f1a705ecea48e30db3c4dc43d6851d) by @mkollo
 ([merge request](texta/texta-rest!162))
- [bug: Fixed issue of indexes not being deleted when deleting a long list of them.](texta/texta-rest@1d1dbc54bbcc9ac2c9aa3aec3071d8e1c61ce742) by @mkollo
 ([merge request](texta/texta-rest!159))

### Updates (3 changes)

- [update: Reformatted TaggerGroups and added tasks for access to timestamps.](texta/texta-rest@cfa20d87a421e86dc5861d8457027579fa1d70ff) by @mkollo
 ([merge request](texta/texta-rest!161))
- [update: Added uuid field to User Profile.](texta/texta-rest@3a515f29f73cd82b44675b2f4fab5a6ddd34cf12) by @mkollo
 ([merge request](texta/texta-rest!160))
- [update: Added missing functionality to remove favourited models.](texta/texta-rest@e95a2f6db7317cd022f39dd052520306cd74094a) by @mkollo
 ([merge request](texta/texta-rest!155))


settings.py
######################### TEXTA Configuration File ########################

# This file contains all the framework's general configurable options which
# don't need changes in the source code. Some configurations are hardcoded
# into different applications' source code and may be altered to fine-tune
# the output or calculation results. For all hardcoded configurations,
# consult #TODO.
# 
# Default configuration suffices for running a development version. For
# production, one needs to define server specific paths.
# 
# General options define file and URL paths, #TODO
# 
# Installation is covered in documentation at #TODO
# 

from time import strftime
import os

# Path to TEXTA's root directory. It is used in other paths as a prefix.
# BASE_DIR = os.path.realpath(os.path.dirname(__file__)) tries to determine
# the path programmatically but may occasionally fail.
# 
BASE_DIR = os.path.realpath(os.path.dirname(__file__))

############################ Server Type ###########################

# Server type allows to predetermine a wide range of options.
# It is used to switch between 'development' and 'production' instances,
# so that after setting up the variables, one only has to alter the
# SERVER_TYPE string to have a fully configured and functional instance. 
#
# The default development version is tuned for hosting TEXTA locally and
# needs no further configuration. However, if your development version is
# on a remote machine, consult 'production' settings.
# 
# URL_PREFIX_DOMAIN - domain address of the hosting server.
# URL_PREFIX_RESOURCE - resource which is responsible for hosting TEXTA
#                       application server.
# ROOT_URLCONF - Django's root urls.py file's package path.
# STATIC_URL - static directory's address.
# STATIC_ROOT - static directory's file path.
# DEBUG - whether to display verbose variable values and stack trace when
#         error occurs during page resolving.

SERVER_TYPE = 'development'

if SERVER_TYPE == 'development':
    URL_PREFIX_DOMAIN = 'http://localhost:8000'
    URL_PREFIX_RESOURCE = ''
    ROOT_URLCONF = 'texta.urls'
    STATIC_URL = URL_PREFIX_DOMAIN + '/static/'
    STATIC_ROOT = os.path.join(BASE_DIR, 'static')
    DEBUG = True
elif SERVER_TYPE == 'production':
    URL_PREFIX_DOMAIN = 'http://textadev.stacc.ee'
    URL_PREFIX_RESOURCE = '/texta'
    ROOT_URLCONF = 'texta.urls'
    STATIC_URL = '/texta/static/'
    DEBUG = False

########################### URLs and paths ###########################

# Defines URLs and paths which are used by various apps.

# URL to TEXTA's root. E.g. 'http://textadev.stacc.ee' or 'http://www.stacc.ee/texta'.
# Should not be altered.
#
URL_PREFIX = URL_PREFIX_DOMAIN + URL_PREFIX_RESOURCE

# URL to the application's login page. TEXTA's login page is its index page.
# Should not be altered. Needed for Django's auth module, used for redirecting
# when unauthorized user is attempting to access authorized content.
#
LOGIN_URL = URL_PREFIX

# Path to media files root directory.
# 
MEDIA_ROOT = os.path.join(BASE_DIR, 'files')

# URL to media files root.
# 
MEDIA_URL = '/files/'
ADMIN_MEDIA_PREFIX = '/media/'

# Path to users' visited words in Lex Miner.
# 
USER_MODELS = os.path.join(BASE_DIR,'data','usermodels')

# Path to users' language models.
# 
MODELS_DIR = os.path.join(BASE_DIR,'data','models')

if not os.path.exists(MODELS_DIR):
    os.makedirs(MODELS_DIR)

if not os.path.exists(MODELS_DIR):
    os.makedirs(MODELS_DIR)

STATICFILES_DIRS = (os.path.join(BASE_DIR, 'static_general'),) # TODO remove


ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

# SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')

# List of all host headers which are accepted to prevent host header poisoning.
# Should be altered if hosted on a remote machine.
#
ALLOWED_HOSTS = ['localhost','texta.stacc.ee','textadev.stacc.ee']

# Defines which database backend does the application use. TEXTA uses only default with sqlite engine.
# Can change engine and database info as one sees fit.
#
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
    # Add 'postgresql_psycopg2', 'postgresql', 'mysql', 'sqlite3' or 'oracle'.
        'NAME': os.path.join(BASE_DIR,'lex.db'),  # Or path to database file if using sqlite3.
        'USER': '',  # Not used with sqlite3.
        'PASSWORD': '',  # Not used with sqlite3.
        'HOST': '',  # Set to empty string for localhost. Not used with sqlite3.
        'PORT': '',  # Set to empty string for default. Not used with sqlite3.
    }
}

TIME_ZONE = 'Europe/Tallinn'
LANGUAGE_CODE = 'et'

SITE_ID = 1
USE_I18N = True
USE_L10N = True

# Key used for seed in various Django's features needing crypto or hashing.
#
SECRET_KEY = '+$18(*8p_h0u6-)z&zu^@=$2h@=8qe+3uwyv+3#v9*)fy9hy&f'

# Defines template engine and context processors to render dynamic HTML pages.
# 
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR,'templates')],
        'APP_DIRS': False,
        'OPTIONS': {
            'context_processors': [
                'django.contrib.auth.context_processors.auth',
                'django.template.context_processors.debug',
                'django.template.context_processors.i18n',
                'django.template.context_processors.media',
                'django.template.context_processors.static',
                'django.template.context_processors.tz',
                'django.template.context_processors.request',
                'django.contrib.messages.context_processors.messages',
                'texta.utils.context_processors.get_version'
            ],
            'loaders': [
                'django.template.loaders.filesystem.Loader',
                'django.template.loaders.app_directories.Loader',
            ],
            'debug': DEBUG,
        },
    },
]

# List of Django plugins used in TEXTA.
# 
MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
)

# List of built-in and custom apps in use.
# 
INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.admin',
    'django.contrib.staticfiles',
    'texta.lm',
    'texta.conceptualiser',
    'texta.mwe_miner',
    'texta.account',
    'texta.home',
    'texta.corpus_tool',
    'texta.model_manager',
    'texta.ontology_viewer',
    'texta.base',
    'texta.permission_admin',
    'texta.grammar_builder',
    'texta.document_miner',
)

############################ Elasticsearch ###########################

# Elasticsearch connection settings. Elasticsearch is used throughout
# TEXTA to store the analyzed documents.
# 


# Elasticsearch URL with protocol specification. Can be either localhost
# or remote address.
#
es_url = os.getenv('TEXTA_ELASTICSEARCH_URL')
if es_url is None:
#    es_url = 'http://localhost:9200'
    es_url = 'http://elasticsearch2.stacc.ee:9201'

# Elasticsearch links to outside world
# ('index_name','mapping_name','field_name'):('url_prefix','url_suffix')
es_links = {
    ('etsa_new', 'event_dgn', 'epiId'): ('https://p12.stacc.ee/common/epicrisis/?id=', ''),
    ('etsa_new', 'event_dgn', 'patId'): ('https://p12.stacc.ee/common/aegread/index.php/aegrida/get/?id=', '')
    }

# Date format used in Elasticsearch fields.
# 
date_format = 'yyyy-MM-dd'

# Set to True if Elasticsearch needs authentication. Tested with basic auth.
es_use_ldap = False
es_ldap_user = os.getenv('TEXTA_LDAP_USER')
es_ldap_password = os.getenv('TEXTA_LDAP_PASSWORD')

############################### Logging ##############################

# TEXTA stores errors and query info in two different log files.
#

# Path to the log directory. Default is /log
# 
LOG_PATH = os.path.join(BASE_DIR,'log')

# Separator used to join different logged features.
#
logging_separator = ' - '

# Paths to info and error log files.
# 
info_log_file_name = os.path.join(LOG_PATH, "info.log")
error_log_file_name = os.path.join(LOG_PATH, "error.log")

# Logger IDs, used in apps. Do not change.
#
INFO_LOGGER = 'info_logger'
ERROR_LOGGER = 'error_logger'

# Most of the following logging settings can be changed.
# Especially format, logging levels, logging class and filenames.
#
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'detailed': {
            'format': logging_separator.join(
                    ['%(levelname)s', '%(module)s', '%(name)s', '%(process)d', '%(thread)d', '%(message)s',
                     '%(asctime)s'])
        },
        'normal': {
            'format': logging_separator.join(['%(levelname)s', '%(module)s', '%(message)s', '%(asctime)s'])
        },
        'detailed_error': {
            'format': '\n' + logging_separator.join(
                    ['%(levelname)s', '%(module)s', '%(name)s', '%(process)d', '%(thread)d', '%(message)s',
                     '%(asctime)s'])
        }
    },
    'handlers': {
        'info_file': {
            'level': 'DEBUG',
            'class': 'logging.FileHandler',
            'formatter': 'detailed',
            'filename': info_log_file_name,
            'encoding': 'utf8',
            'mode': 'a'
        },
        'error_file': {
            'level': 'ERROR',
            'class': 'logging.FileHandler',
            'formatter': 'detailed_error',
            'filename': error_log_file_name,
            'encoding': 'utf8',
            'mode': 'a'
        },
    },
    'loggers': {
        INFO_LOGGER: {
            'level': 'DEBUG',
            'handlers': ['info_file']
        },
        ERROR_LOGGER: {
            'level': 'ERROR',
            'handlers': ['error_file']
        }
    }
}

############################ Boot scripts ###########################

# Several scripts ran during the boot to set up files and directories

from texta.utils.setup import write_navigation_file, ensure_dir_existence

write_navigation_file(URL_PREFIX, STATIC_URL)
ensure_dir_existence(LOG_PATH)
ensure_dir_existence(MODELS_DIR)
ensure_dir_existence(USER_MODELS)